### PR TITLE
GUI: create new datasets

### LINKED ISSUE #1707: Dashboard: create dataset
# ARTIQ Feature Request

## Problem this request addresses

Currently it's easy to edit / delete a dataset from the dashboard but not to create a new one. Is there a reason for this, or just an omission?

## Describe the solution you'd like

Option to create a new dataset from the dashboard.

### PR DESCRIPTION
**Pull request by SidaChen1999**
# ARTIQ Pull Request

## Description of Changes
Adding new datasets function in artiq_dashboard. The users can type in the name and value of the datasets in corresponding boxes. The data type will be shown next to the value box in real-time. It will show "N/A" if the input cannot be recognized and if the user still hits enter it will not proceed but show a log error. The decoder is `sipyco::pyon`
Warn!: string should be input like `"something"` with a double-quote.

image

### Related Issue

Closes #1707 

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :sparkles: New feature |

### All Pull Requests

- [x] Use correct spelling and grammar.
- [x] Close/update issues.

### Code Changes

- [x] Run `flake8` to check code style (follow PEP-8 style). `flake8` has issues with parsing Migen/gateware code, ignore as necessary.
- [ ] Test your changes or have someone test them. Mention what was tested and how.

### Git Logistics

- [x] Split your contribution into logically separate changes (`git rebase --interactive`). Merge/squash/fixup commits that just fix or amend previous commits. Remove unintended changes & cleanup. See [tutorial](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase).
- [x] Write short & meaningful commit messages. Review each commit for messages (`git show`). Format:
  ```
  topic: description. < 50 characters total.
  
  Longer description. < 70 characters per line
  ```

### Licensing

See [copyright & licensing for more info](https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#copyright-and-sign-off).
ARTIQ files that do not contain a license header are copyrighted by M-Labs Limited and are licensed under LGPLv3+.

### CODE REVIEW

**Comment by dnadlinger** [comment]
The "real-time" data type display is a neat touch; I like it.

**Comment by hartytp** [comment]
❤️

File: artiq/dashboard/datasets.py [hunk 1]
```diff
@@ -81,6 +82,63 @@ def get_edit_widget(self, initial_value):
...
+        self.dataset_ctl = dataset_ctl
+
+        self.setWindowTitle("Create dataset")
+        grid = QtWidgets.QGridLayout()
+        self.setLayout(grid)
+
+        grid.addWidget(QtWidgets.QLabel("Name:"), 0, 0)
+        grid.addWidget(self.get_key_widget(), 0, 1)
+
+        grid.addWidget(QtWidgets.QLabel("Value:"), 1, 0)
```
Comment (dnadlinger) [inline artiq/dashboard/datasets.py:97]: Maybe add `" (PYON):"` to the label, or at least as a tool tip?

File: artiq/dashboard/datasets.py [hunk 2]
```diff
@@ -81,6 +82,63 @@ def get_edit_widget(self, initial_value):
...
+
+    def accept(self):
+        try:
+            key = self.get_create_widget_text()
+            value = self.get_create_widget_value()
+            asyncio.ensure_future(self.dataset_ctl.set(key, pyon.decode(value)))
+            QtWidgets.QDialog.accept(self)
+        except:
+            logger.error("Cannot create dataset: "
+                                 "type is not supported",)
```
Comment (dnadlinger) [inline artiq/dashboard/datasets.py:118]: Indentation weird. Also, isn't "type not supported" potentially a bit misleading for what is most likely a user typo?

File: artiq/dashboard/datasets.py [hunk 3]
```diff
@@ -81,6 +82,63 @@ def get_edit_widget(self, initial_value):
...
+            return
+
+    def get_key_widget(self):
+        self.key_widget = QtWidgets.QLineEdit()
```
Comment (dnadlinger) [inline artiq/dashboard/datasets.py:122]: I find a `get_*` method that overwrites a member to be rather confusing. I'd either just use `self.key_widget` throughout, or make the `get_*` methods lazily create the line edits (i.e. something like `if self.key_widget is None: … return self.key_widget`).

**Review by dnadlinger — CHANGES_REQUESTED** [review]
Perhaps the most intuitive UI choice would be to highlight an invalid value (or non-allowable key) somehow, and disallow form submission then. A red border on the line edit, perhaps. This would make the UI rather self-documenting.

The UI should also have a way to specify whether the dataset is persistent (probably just a tick box).

I've left a few inline code comments as well.

**Comment by dhslichter** [comment]
this is fantastic @SidaChen1999!  I agree with @dnadlinger's comments, as well as his suggestion for a "persist" tickbox.

**Comment by sbourdeauducq** [comment]
> It will show "N/A" if the input cannot be recognized and if the user still hits enter it will not proceed but show a log error.

What about greying out the OK button and ignoring Enter while the input doesn't parse?

**Comment by SidaChen1999** [comment]
Thanks for all the reviews. Now the window will look like the first picture when it is first opened. And when something invalid is input, it will show as the second one.
image
image

File: .gitignore [hunk 4]
```diff
@@ -16,7 +16,7 @@ __pycache__/
 /.coverage
 /artiq/test/lit/*/Output/
 /artiq/binaries
-/artiq/firmware/target/
+/artiq/firmware/**/target/
```
Comment (sbourdeauducq) [inline .gitignore:19]: This change does not belong in this PR.

File: artiq/dashboard/datasets.py [hunk 5]
```diff
@@ -81,6 +82,89 @@ def get_edit_widget(self, initial_value):
...
+        grid.addWidget(self.buttons, 4, 0, 1, 3)
+        self.buttons.accepted.connect(self.accept)
+        self.buttons.rejected.connect(self.reject)
+
+    def accept(self):
+        try:
+            key = self.get_create_widget_text()
+            value = self.get_create_widget_value()
+            persist = self.get_persist_widget_value()
+            asyncio.ensure_future(self.dataset_ctl.set(key, pyon.decode(value), persist))
```
Comment (sbourdeauducq) [inline artiq/dashboard/datasets.py:124]: Use this to wrap the asyncio task and log errors immediately: https://github.com/m-labs/artiq/blob/master/artiq/tools.py#L107
Otherwise the asyncio task exceptions are only retrieved when the GUI exits and error messages printed then. This makes debugging tedious.

File: artiq/dashboard/datasets.py [hunk 6]
```diff
@@ -81,6 +82,89 @@ def get_edit_widget(self, initial_value):
...
+        return self.box_widget
+
+    def get_create_widget_text(self):
+        return self.name_widget.text()
+
+    def get_create_widget_value(self):
+        return self.value_widget.text()
+    
+    def get_persist_widget_value(self):
+        return self.box_widget.isChecked()
```
Comment (sbourdeauducq) [inline artiq/dashboard/datasets.py:152]: Those methods can be removed IMO. The ones that create the widgets are also not very well-written.

File: artiq/dashboard/datasets.py [hunk 7]
```diff
@@ -81,6 +82,89 @@ def get_edit_widget(self, initial_value):
...
+        except:
+            self.value_widget.setStyleSheet("border: 1px solid red;"
+                                            "height: 28px;"
+                                            "border-radius: 2px;")
+            self.data_type.setText('N/A')
+            self.ok.setEnabled(False)
+        else:
+            self.value_widget.setStyleSheet("")
+            self.ok.setEnabled(True)
+        return
```
Comment (sbourdeauducq) [inline artiq/dashboard/datasets.py:167]: This ``return`` is not needed, remove.

File: artiq/dashboard/datasets.py [hunk 8]
```diff
@@ -146,6 +235,10 @@ def set_model(self, model):
         self.table_model_filter.setSourceModel(self.table_model)
         self.table.setModel(self.table_model_filter)
 
+    def create_clicked(self):
+        dialog_cls = Creator
+        dialog_cls(self, self.dataset_ctl).open()
```
Comment (sbourdeauducq) [inline artiq/dashboard/datasets.py:240]: The ``dialog_cls`` variable is not needed and should be removed.

File: artiq/dashboard/datasets.py [hunk 9]
```diff
@@ -81,6 +82,89 @@ def get_edit_widget(self, initial_value):
...
+        self.buttons.rejected.connect(self.reject)
+
+    def accept(self):
+        try:
+            key = self.get_create_widget_text()
+            value = self.get_create_widget_value()
+            persist = self.get_persist_widget_value()
+            asyncio.ensure_future(self.dataset_ctl.set(key, pyon.decode(value), persist))
+            QtWidgets.QDialog.accept(self)
+        except:
```
Comment (sbourdeauducq) [inline artiq/dashboard/datasets.py:126]: Remove.

File: artiq/dashboard/datasets.py [hunk 10]
```diff
@@ -81,6 +82,89 @@ def get_edit_widget(self, initial_value):
...
+            logger.error("Cannot create dataset: "
+                         "input is not valid",)
+            return
```
Comment (sbourdeauducq) [inline artiq/dashboard/datasets.py:129]: Remove.

File: artiq/dashboard/datasets.py [hunk 11]
```diff
@@ -82,6 +83,69 @@ def get_edit_widget_value(self):
...
+            key, pyon.decode(value), persist)))
+        QtWidgets.QDialog.accept(self)
+
+    def dtype(self):
+        txt = self.value_widget.text()
+        try:
+            self.data_type.setText(type(pyon.decode(txt)).__name__)
+        except:
+            self.value_widget.setStyleSheet("border: 1px solid red;"
+                                            "height: 28px;"
```
Comment (dnadlinger) [inline artiq/dashboard/datasets.py:140]: Why the explicit `height`? Is this compatible across OSes/UI toolkits?

Comment (SidaChen1999) [reply artiq/dashboard/datasets.py:140]: I tested on KDE and XFce, style sheet of `border: 1px solid red` will change the height differently from the default. I fixed the `height` to be the same in KDE but it is indeed different from XFce. So I'll look for another solution.

Comment (SidaChen1999) [reply artiq/dashboard/datasets.py:140]: I also tested on GNOME. Changing the border properties of the style sheet will also affect padding and some other properties that make it inconsistent and unaesthetic across OSes. Therefore, I propose to change the background color of the line edit box instead as shown below
image

Comment (sbourdeauducq) [reply artiq/dashboard/datasets.py]: I'm not sure if a hardcoded color is a good idea, since the user-defined theme colors may change. One obvious problem: You are using the dark theme, but most people use the default with black text and it will be hard to read on a red background.
What about displaying a warning icon at the right of the text box (possibly inside it) when the input is incorrect? Or instead of the N/A text?

Comment (SidaChen1999) [reply artiq/dashboard/datasets.py]: I changed the `N/A` text into `MessageBoxWarning`. The icon and its size will change according to the OS.
image

File: artiq/dashboard/datasets.py [hunk 12]
```diff
@@ -82,6 +83,67 @@ def get_edit_widget_value(self):
...
+        persist = self.box_widget.isChecked()
+        asyncio.ensure_future(exc_to_warning(self.dataset_ctl.set(
+            key, pyon.decode(value), persist)))
+        QtWidgets.QDialog.accept(self)
+
+    def dtype(self):
+        txt = self.value_widget.text()
+        try:
+            self.data_type.setText(type(pyon.decode(txt)).__name__)
+            self.value_widget.setStyleSheet("")
```
Comment (sbourdeauducq) [inline artiq/dashboard/datasets.py:138]: remove

File: artiq/dashboard/datasets.py [hunk 13]
```diff
@@ -82,6 +83,67 @@ def get_edit_widget_value(self):
...
+        value = self.value_widget.text()
+        persist = self.box_widget.isChecked()
+        asyncio.ensure_future(exc_to_warning(self.dataset_ctl.set(
+            key, pyon.decode(value), persist)))
+        QtWidgets.QDialog.accept(self)
+
+    def dtype(self):
+        txt = self.value_widget.text()
+        try:
+            self.data_type.setText(type(pyon.decode(txt)).__name__)
```
Comment (sbourdeauducq) [inline artiq/dashboard/datasets.py:137]: limit the try/except to ``pyon.decode``

Comment (SidaChen1999) [reply artiq/dashboard/datasets.py:137]: I think it is already limited by the `textChanged` signal. Shouldn't we validate the input as soon as new input is detected? I tested on the old pc and it won't take too much resource.

Comment (sbourdeauducq) [reply artiq/dashboard/datasets.py:137]: What I meant is right now you are catching exceptions raised by e.g. ``setText`` or ``setEnabled``. This has nothing to do with resource consumption.

File: artiq/dashboard/datasets.py [hunk 14]
```diff
@@ -134,14 +134,15 @@ def accept(self):
     def dtype(self):
         txt = self.value_widget.text()
         try:
-            self.data_type.setText(type(pyon.decode(txt)).__name__)
-            self.value_widget.setStyleSheet("")
-            self.ok.setEnabled(True)
+            pyon.decode(txt)
```
Comment (sbourdeauducq) [inline artiq/dashboard/datasets.py:137]: Just store the result and decode once. You don't need to do anything special regarding try/else:
```python
>>> try:
...    x = 2
... except:
...    pass
... else:
...    print(x)
... 
2
```

Please do such research on the Python language next time.
